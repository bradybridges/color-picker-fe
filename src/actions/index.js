export const addProjects = (projects) => ({
  type: 'ADD_PROJECTS',
  projects,
});

export const addPalettes = (palettes) => ({
  type: 'ADD_PALETTES',
  palettes,
});
