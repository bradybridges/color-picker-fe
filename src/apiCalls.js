export const getProjects = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/projects');
  if(!response.ok) {
    return Error('Failed to fetch projects');
  }
  const projects = await response.json();
  return projects;
};

export const getPalettes = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes');
  if(!response.ok) {
    return Error('Failed to fetch palettes');
  }
  const palettes = await response.json();
  return palettes;
};

export const getProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`);
  if(!response.ok) {
    return Error('Failed to fetch pallete');
  }
  const palette = await response.json();
  return palette;
};

export const getPalette = async (paletteId) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${paletteId}`);
  if (!response.ok) {
    return Error('Failed to fetch palette');
  }
  const palette = await response.json();
  return palette;
};

export const postProject = async (name) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  if (!response.ok) {
    return Error('Failed to post new project');
  }
  const message = await response.json();
  console.log('message', message)
  return message;
};

export const postPalette = async (project_id, palette_name, color_1, color_2, color_3, color_4, color_5) => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(
      {
        project_id,
        palette_name,
        color_1,
        color_2,
        color_3,
        color_4,
        color_5,
      },
    ),
  });
  if (!response.ok) {
    return response;
  }
  const message = await response.json();
  return message;
};

export const deleteProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete project');
  }
  const message = await response.json();
  return message;
};

export const deletePalette = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete palette');
  }
  const message = await response.json();
  return message;
};

export const patchProject = async () => {

};

export const patchPalette = async () => {

};
