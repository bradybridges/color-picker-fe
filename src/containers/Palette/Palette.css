.palette {
  /* background-color: white; */
  padding: 2px;
  width: 90%;
  justify-content: center;
  border: 1px solid white;
  margin: 4px;
}

.title-container {
  display: flex;
  justify-content: space-between;
  margin-bottom: 5px;
}

.palette-name {
  margin: 5px;
}

.delete-button { 
  width: 32px;
  height: 32px;
  margin: 0;
}

.palette-color-container {
  display: flex;
  height: 30px;
  justify-content: center;
}

.color1,
.color2,
.color3,
.color4,
.color5 {
  height: 100%;
  width: 20%;
  margin: 1px;
}