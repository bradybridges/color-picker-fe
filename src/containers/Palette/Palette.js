import React, { Component } from 'react'
import './Palette.css'
import { connect } from 'react-redux';
import { addPalettes, setTempPalette } from '../../actions/index';
import { deletePalette } from '../../apiCalls';
const deleteButton = require('../../images/delete.svg');
export class Palette extends Component {
  handleDeletePalette = async () => {
    const id = Number(this.props.id);
    const updatedPalettes = this.props.palettes.filter((palette) => palette.id !== id);
    this.props.updatePalettes(updatedPalettes);
    const response = await deletePalette(id);
    console.log(response);
  }

  loadPalette = () => {
    const { color_1, color_2, color_3, color_4, color_5 } = this.props;
    const tempPalette = [
      { name: "color_1", color: color_1, isLocked: false, },
      { name: "color_2", color: color_2, isLocked: false, },
      { name: "color_3", color: color_3, isLocked: false, },
      { name: "color_4", color: color_4, isLocked: false, },
      { name: "color_5", color: color_5, isLocked: false, },
    ];
    this.props.setTempPalette(tempPalette);
  }

  render() {
    const { palette_name, color_1, color_2, color_3, color_4, color_5 } = this.props;
    return (
      <section className='palette' onClick={this.loadPalette}>
        <div className='title-container'>
          <h6 className='palette-name'>{palette_name}</h6>
          <img 
            className='delete-button' 
            src={deleteButton} 
            alt="delete button" 
            onClick={this.handleDeletePalette} 
          />
        </div>
        <div className='palette-color-container'>
          <div className='color1' style={{backgroundColor: color_1}} />
          <div className='color2' style={{backgroundColor: color_2}} />
          <div className='color3' style={{backgroundColor: color_3}} />
          <div className='color4' style={{backgroundColor: color_4}} />
          <div className='color5' style={{backgroundColor: color_5}} />
        </div>
      </section>
    );
  }
}

export const mapState = (state) => ({
  palettes: state.palettes,
});

export const mapDispatch = (dispatch) => ({
  updatePalettes: (palettes) => dispatch(addPalettes(palettes)),
  setTempPalette: (palette) => dispatch(setTempPalette(palette)),
});

export default connect(mapState, mapDispatch)(Palette);
