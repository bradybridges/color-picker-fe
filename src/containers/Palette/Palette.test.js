import React from 'react';
import { shallow } from 'enzyme';
import { Palette, mapState, mapDispatch } from './Palette';
import * as actions from '../../actions/index';
import * as api from '../../apiCalls';

describe('Palette', () => {
  const mockPalettes = [
    {
      id: 1,
      project_id: 1,
      palette_name: 'super sad palette',
      color_1: '#999999',
      color_2: '#1c1c1c',
      color_3: '#1d1d1d',
      color_4: '#1f1f1f',
      color_5: '#1a1a1a',
    },
    {
      id: 2,
      project_id: 1,
      palette_name: 'happy palette',
      color_1: '#FFFFFF',
      color_2: '#000000',
      color_3: '#CCCCCC',
      color_4: '#1F1F1F',
      color_5: '#1D1D1D',
    },
  ];
  const mockUpdatePalettes = jest.fn();
  const mockSetTempPalette = jest.fn();
  const name = 'New Palette';
  const color1 = '#000000';
  const color2 = '#FFFFFF';
  const color3 = '#CCCCCC';
  const color4 = '#1F1F1F';
  const color5 = '#1D1D1D';
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Palette
        id={1}
        palette_name={name}
        color_1={color1}
        color_2={color2}
        color_3={color3}
        color_4={color4}
        color_5={color5}
        palettes={mockPalettes}
        updatePalettes={mockUpdatePalettes}
        setTempPalette={mockSetTempPalette}
      />
    );
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('handleDeletePalette should update palettes in store and delete palette from BE', () => {
    const expected = [
      {
        id: 2,
        project_id: 1,
        palette_name: 'happy palette',
        color_1: '#FFFFFF',
        color_2: '#000000',
        color_3: '#CCCCCC',
        color_4: '#1F1F1F',
        color_5: '#1D1D1D',
      },
    ];
    api.deletePalette = jest.fn();
    wrapper.instance().handleDeletePalette();
    expect(mockUpdatePalettes).toHaveBeenCalledWith(expected);
    expect(api.deletePalette).toHaveBeenCalledWith(1);
  });
  it('loadPalette should set tempPalette with correct colors', () => {
    const expected = [
      { color: '#000000', isLocked: false, name: 'color_1' },
      { color: '#FFFFFF', isLocked: false, name: 'color_2' },
      { color: '#CCCCCC', isLocked: false, name: 'color_3' },
      { color: '#1F1F1F', isLocked: false, name: 'color_4' },
      { color: '#1D1D1D', isLocked: false, name: 'color_5' },
    ];
    wrapper.instance().loadPalette();
    expect(mockSetTempPalette).toHaveBeenCalledWith(expected);
  });
});

describe('mapState', () => {
  const mockPalettes = [
    {
      id: 1,
      project_id: 1,
      palette_name: 'super sad palette',
      color_1: '#999999',
      color_2: '#1c1c1c',
      color_3: '#1d1d1d',
      color_4: '#1f1f1f',
      color_5: '#1a1a1a',
    },
    {
      id: 2,
      project_id: 1,
      palette_name: 'happy palette',
      color_1: '#FFFFFF',
      color_2: '#000000',
      color_3: '#CCCCCC',
      color_4: '#1F1F1F',
      color_5: '#1D1D1D',
    },
  ];
  it('should map palettes to state', () => {
    const mockState = {
      error: 'Something outrageous happened',
      palettes: mockPalettes,
    };
    const expected = { palettes: mockPalettes };
    const mappedState = mapState(mockState);
    expect(mappedState).toEqual(expected);
  });
});

describe('mapDispatch', () => {
  it('setTempPalettes should be dispatched with the correct action obj', () => {
    const mockDispatch = jest.fn();
    const mockTemp = [
      {
        name: 'color_1',
        color: '#FFFFFF',
        isLocked: false,
      },
      {
        name: 'color_2',
        color: '#000000',
        isLocked: false,
      },
      {
        name: 'color_3',
        color: '#CCCCCC',
        isLocked: false,
      },
      {
        name: 'color_4',
        color: '#1F1F1F',
        isLocked: true,
      },
      {
        name: 'color_5',
        color: '#1D1D1D',
        isLocked: true,
      },
    ];
    const actionToDispatch = actions.setTempPalette(mockTemp);
    const mappedDispatch = mapDispatch(mockDispatch);
    mappedDispatch.setTempPalette(mockTemp);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('updatePalettes should be called with the correct action obj', () => {
    const mockDispatch = jest.fn();
    const mockPalettes = [
      {
        id: 1,
        palette_name: 'Great Palette',
        color_1: '#000000',
        color_2: '#CCCCCC',
        color_3: '#FFFFFF',
        color_4: '#1F1F1F',
        color_5: '#1D1D1D',
      },
      {
        id: 2,
        palette_name: 'Sweet Palette',
        color_1: '#010101',
        color_2: '#C2C2C2',
        color_3: '#FFFFFF',
        color_4: '#1F1F1F',
        color_5: '#AF1D1D',
      },
    ];
    const actionToDispatch = actions.addPalettes(mockPalettes);
    const mappedDispatch = mapDispatch(mockDispatch);
    mappedDispatch.updatePalettes(mockPalettes);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});
