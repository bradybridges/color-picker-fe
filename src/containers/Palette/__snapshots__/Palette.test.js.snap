// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette should match snapshot 1`] = `
<section
  className="palette"
  onClick={[Function]}
>
  <div
    className="title-container"
  >
    <h6
      className="palette-name"
    >
      New Palette
    </h6>
    <img
      alt="delete button"
      className="delete-button"
      onClick={[Function]}
      src={
        Object {
          "ReactComponent": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
          "__esModule": true,
          "default": "delete.svg",
        }
      }
    />
  </div>
  <div
    className="palette-color-container"
  >
    <div
      className="color1"
      style={
        Object {
          "backgroundColor": "#000000",
        }
      }
    />
    <div
      className="color2"
      style={
        Object {
          "backgroundColor": "#FFFFFF",
        }
      }
    />
    <div
      className="color3"
      style={
        Object {
          "backgroundColor": "#CCCCCC",
        }
      }
    />
    <div
      className="color4"
      style={
        Object {
          "backgroundColor": "#1F1F1F",
        }
      }
    />
    <div
      className="color5"
      style={
        Object {
          "backgroundColor": "#1D1D1D",
        }
      }
    />
  </div>
</section>
`;
